File: core/src/main/java/ru/hits/core/domain/dto/operation/OperationDTO.java
package ru.hits.core.domain.dto.operation;

import io.swagger.v3.oas.annotations.media.Schema;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import ru.hits.core.domain.enums.OperationTypeEnum;

import java.time.Instant;
import java.util.UUID;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Schema(description = "DTO, представляющий операцию со счётом")
public class OperationDTO {

    @Schema(description = "Уникальный идентификатор операции", example = "a1234567-b89b-12d3-a456-426614174000")
    private UUID id;

    @Schema(description = "Идентификатор счёта отправителя", example = "c1234567-d89b-12d3-a456-426614174000")
    private UUID senderAccountId;

    @Schema(description = "Номер счёта отправителя", example = "1234 1234 1324 1324")
    private String senderAccountNumber;

    @Schema(description = "Идентификатор счёта получателя", example = "d1234567-e89b-12d3-a456-426614174000")
    private UUID recipientAccountId;

    @Schema(description = "Номер счёта отправителя", example = "1234 1234 1324 1324")
    private String recipientAccountNumber;

    @Schema(description = "Значение определяющее была ли операция плюсом к счету или минусом", example = "true")
    private Boolean directionToMe;

    @Schema(description = "Сумма операции", example = "500.75")
    private Float amount;

    @Schema(description = "Дата и время операции", example = "2024-02-22T14:30:00Z")
    private Instant transactionDateTime;

    @Schema(description = "Сообщение, сопровождающее операцию", example = "Оплата за услуги")
    private String message;

    @Schema(description = "Тип операции", example = "replenishment")
    private OperationTypeEnum operationType;

    @Schema(description = "Флаг просроченной оплаты кредита", example = "false")
    private Boolean isPaymentExpired;
}

File: employee-app/HbankPRO/app/src/main/java/com/example/h_bankpro/data/dto/OperationDto.kt
package com.example.h_bankpro.data.dto

import com.example.h_bankpro.data.OperationType
import com.example.h_bankpro.domain.model.Operation
import kotlinx.datetime.Instant
import kotlinx.datetime.toJavaInstant
import kotlinx.serialization.Serializable
import java.time.ZoneId

@Serializable
data class OperationDto(
    val id: String,
    val senderAccountId: String?,
    val senderAccountNumber: String?,
    val recipientAccountId: String?,
    val recipientAccountNumber: String?,
    val directionToMe: Boolean,
    val amount: Double,
    val transactionDateTime: Instant,
    val message: String?,
    val operationType: OperationType,
    val isPaymentExpired: Boolean?
)

internal fun OperationDto.toDomain(): Operation =
    Operation(
        id = id,
        senderAccountId = senderAccountId,
        senderAccountNumber = senderAccountNumber,
        recipientAccountId = recipientAccountId,
        recipientAccountNumber = recipientAccountNumber,
        directionToMe = directionToMe,
        amount = amount,
        transactionDateTime = transactionDateTime.toJavaInstant()
            .atZone(ZoneId.systemDefault())
            .toLocalDateTime(),
        message = message,
        operationType = operationType,
        isPaymentExpired = isPaymentExpired
    )